### PR TITLE
feat(virtio): Adding Virtio-net Support to Cloudlet

### LINKED ISSUE #3: Add virtio-net support to Lumper

### PR DESCRIPTION
**Pull request by sylvain-pierrot**
# Adding Virtio-net Support to Cloudlet

**Description**:

This Pull Request aims to enhance the functionality of Cloudlet by adding support for virtio-net, a paravirtualized network device.

**Changes Introduced (checklist)**:

Virtio Over MMIO:

- [x] MMIO Device Discovery
  - MMIO provides no generic device discovery mechanism. Under Linux, discovery is performed via the kernel command line. Ref `VMMIO` from [doc](https://www.kernel.org/doc/html/v4.14/admin-guide/kernel-parameters.html).
- [x] MMIO Device Register Layout
  - MMIO virtio devices provide a set of memory mapped control registers followed by a device-specific configuration space. Ref [here](https://docs.oasis-open.org/virtio/virtio/v1.3/csd01/virtio-v1.3-csd01.html#x1-1820002).
  
 Related Issues:
 
 #3

### CODE REVIEW

File: src/vmm/src/core/devices/virtio/net/device.rs [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+
+pub struct Net {
+    vmmio_parameter: String,
+}
+
+impl Net {
+    pub fn new(
+        size: GuestUsize,
+        baseaddr: GuestAddress,
+        irq: u32,
```
Comment (Assstra) [inline src/vmm/src/core/devices/virtio/net/device.rs:47]: missing device id ?

File: src/vmm/Cargo.toml [hunk 2]
```diff
@@ -23,6 +23,12 @@ openpty = "0.2.0"
 tracing = "0.1.40"
 tracing-subscriber = "0.3.18"
 virtio-bindings = "0.2.2"
+virtio-blk = { git = "https://github.com/rust-vmm/vm-virtio.git", features = [
```
Comment (Assstra) [inline src/vmm/Cargo.toml]: Why do you need virtio block device ?

Comment (Assstra) [reply src/vmm/Cargo.toml]: ?

Comment (sylvain-pierrot) [reply src/vmm/Cargo.toml]: I removed it

File: src/vmm/src/core/devices/virtio/net/device.rs (hunk 1 above)
Comment (sylvain-pierrot) [reply src/vmm/src/core/devices/virtio/net/device.rs:47]: `id` is of type option. For now, `id` is set to None

File: src/vmm/src/core/vmm.rs [hunk 3]
```diff
@@ -277,15 +340,63 @@ impl VMM {
...
+            .map_err(|_| Error::MmioRange)?;
+        let irq = self.irq_allocator.next_irq().map_err(Error::IrqAllocator)?;
+        let mmio_cfg = MmioConfig {
+            range: mmio_range,
+            gsi: irq,
+        };
+        // let mut guard = self.device_mgr.lock().unwrap();
+
+        // !TODO: MMIO Device Discovery + MMIO Device Register Layout
+        let net = Net::new(
```
Comment (Kuruyia) [inline src/vmm/src/core/vmm.rs]: Those can be removed.
```suggestion
        };

        let net = Net::new(
```

File: src/vmm/src/core/vmm.rs [hunk 4]
```diff
@@ -221,6 +273,17 @@ impl VMM {
...
 
+        let event_mgr = self.event_mgr.clone();
+        let _ = thread::Builder::new().spawn(move || loop {
+            match event_mgr.lock().unwrap().run() {
+                Ok(_) => (),
+                Err(e) => eprintln!("Failed to handle events: {:?}", e),
+            }
+            // if !self.exit_handler.keep_running() {
+            //     break;
+            // }
```
Comment (Kuruyia) [inline src/vmm/src/core/vmm.rs]: This can be removed.
```suggestion
```

File: src/vmm/src/core/devices/virtio/register.rs [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+use vm_memory::{Address, GuestAddress, GuestUsize};
+
+pub fn register_mmio_device(
+    mmio_cfg: MmioConfig,
+    device_mgr: Arc<Mutex<IoManager>>,
+    irq: u32,
+    id: Option<u32>,
+    device: Arc<dyn DeviceMmio + Send + Sync>,
+) -> Result<String> {
+    // !TODO Register to MmioManager
```
Comment (Kuruyia) [inline src/vmm/src/core/devices/virtio/register.rs]: This can be removed.

```suggestion
```

**Review by Kuruyia — CHANGES_REQUESTED** [review]

File: src/vmm/src/core/vmm.rs (hunk 4 above)
Comment (sylvain-pierrot) [reply src/vmm/src/core/vmm.rs]: Done

File: src/vmm/src/core/devices/virtio/register.rs (hunk 5 above)
Comment (sylvain-pierrot) [reply src/vmm/src/core/devices/virtio/register.rs]: Done

File: src/vmm/src/core/vmm.rs (hunk 3 above)
Comment (sylvain-pierrot) [reply src/vmm/src/core/vmm.rs]: Done

**Review by Kuruyia — APPROVED** [review]
LGTM, nice work!

**Pull request merged**